### ISSUE TITLE
Python auto-instrumentation is not working.

### ISSUE THREAD

**Issue by sang-w0o** [open-telemetry/opentelemetry-operator#1468]
Hello, I encountered an issue while following documentation about [K8S Operator](https://opentelemetry.io/docs/k8s-operator/).

I have installed K8S Operator, and created `OpenTelemetryCollector` and `Instrumentation`.

```yaml
apiVersion: opentelemetry.io/v1alpha1
kind: OpenTelemetryCollector
metadata:
  name: otm-poc-collector
spec:
  config: |
    receivers:
      otlp:
        protocols:
          grpc:
          http:
    processors:
      memory_limiter:
        check_interval: 1s
        limit_percentage: 75
        spike_limit_percentage: 15
      batch:
        send_batch_size: 10000
        timeout: 10s

    exporters:
      otlp:
        endpoint: tempo-us-central1.grafana.net:443 
        headers:
          authorization: Basic <base64-encoded-credentials>

    service:
      pipelines:
        traces:
          receivers: [otlp]
          processors: [memory_limiter, batch]
          exporters: [otlp]
        metrics:
          receivers: [otlp]
          processors: [memory_limiter, batch]
          exporters: [otlp]
        logs:
          receivers: [otlp]
          processors: [memory_limiter, batch]
          exporters: [otlp]

---
apiVersion: opentelemetry.io/v1alpha1
kind: Instrumentation
metadata:
  name: otm-poc-instrumentation
spec:
  exporter:
    endpoint: http://otm-poc-collector-collector:4318
  propagators:
    - tracecontext
    - baggage
  sampler:
    type: parentbased_traceidratio
    argument: "1"
```

When I apply the [required annotation](https://opentelemetry.io/docs/k8s-operator/automatic/#add-annotations-to-existing-deployments) to my pod spec to opt-in to auto-instrumentation, my pod crashes with errors below.

```Defaulted container "identity" out of: identity, opentelemetry-auto-instrumentation (init)
Using default sampler. Failed to initialize custom sampler, parentbased_traceidratio: Requested component 'parentbased_traceidratio' not found in entry points for 'opentelemetry_traces_sampler'
Traceback (most recent call last):
  File "/usr/local/bin/spaceone", line 33, in <module>
    sys.exit(load_entry_point('spaceone-core==1.11.1.dev1', 'console_scripts', 'spaceone')())
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1130, in __call__
    return self.main(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1055, in main
    rv = self.invoke(ctx)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1657, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 1404, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/usr/local/lib/python3.8/site-packages/click/core.py", line 760, in invoke
    return __callback(*args, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/spaceone/core/command.py", line 37, in grpc
    pygrpc.serve()
  File "/usr/local/lib/python3.8/site-packages/spaceone/core/pygrpc/server.py", line 105, in serve
    server = grpc.server(
  File "/otel-auto-instrumentation/opentelemetry/instrumentation/grpc/__init__.py", line 338, in server
    kwargs["interceptors"].insert(
AttributeError: 'tuple' object has no attribute 'insert'
```

python version is 3.8.

Thanks in advance ✋🏻

**Comment by TylerHelmuth** [comment]
What version of the operator are you using? What packages does your application use that can be auto-instrumented?

**Comment by sang-w0o** [comment]
> What version of the operator are you using? What packages does your application use that can be auto-instrumented?

I am using custom framework to serve gRPC requests in my python application.

Below is the output of `kubectl get <otel-collector-pod-id> -o yaml`.

```
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/psp: eks.privileged
    opentelemetry-operator-config/sha256: 740ece111239c828b1acdb070db0b473c4ebe2eca5ec93882ff416b588079d1f
  creationTimestamp: "2023-02-13T01:17:13Z"
  generateName: otel-poc-collector-764cd79d9c-
  labels:
    app.kubernetes.io/component: opentelemetry-collector
    app.kubernetes.io/instance: otel.otel-poc
    app.kubernetes.io/managed-by: opentelemetry-operator
    app.kubernetes.io/name: otel-poc-collector
    app.kubernetes.io/part-of: opentelemetry
    app.kubernetes.io/version: latest
    pod-template-hash: 764cd79d9c
  name: otel-poc-collector-764cd79d9c-k64nl
  namespace: otel
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: otel-poc-collector-764cd79d9c
    uid: 1b3ce004-cc95-4bdd-a087-2763288bf7c4
  resourceVersion: "9108199"
  uid: dc84c36b-0e3f-44bd-8b1f-9118f7c31390
spec:
  containers:
  - args:
    - --config=/conf/collector.yaml
    env:
    - name: POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    image: ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector:0.69.0
    imagePullPolicy: IfNotPresent
    name: otc-container
    ports:
    - containerPort: 8888
      name: metrics
      protocol: TCP
    - containerPort: 4317
      name: otlp-grpc
      protocol: TCP
    - containerPort: 4318
      name: otlp-http
      protocol: TCP
    - containerPort: 55681
      name: otlp-http-legac
      protocol: TCP
    resources: {}
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /conf
      name: otc-internal
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-gx8zc
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  nodeName: #..
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext: {}
  serviceAccount: otel-poc-collector
  serviceAccountName: otel-poc-collector
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - configMap:
      defaultMode: 420
      items:
      - key: collector.yaml
        path: collector.yaml
      name: otel-poc-collector
    name: otc-internal
  - name: kube-api-access-gx8zc
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2023-02-13T01:17:13Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2023-02-13T01:17:15Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2023-02-13T01:17:15Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2023-02-13T01:17:13Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: docker://ff8be38aa83ac757fed99835e783bae8d1120a1d49110df75f9bfbbc9a35d804
    image: ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector:0.69.0
    imageID: docker-pullable://ghcr.io/open-telemetry/opentelemetry-collector-releases/opentelemetry-collector@sha256:6150b0c7226a81ed659bfeb8ea12b3be120e081c778addcb3a4b1e5444a31e1f
    lastState: {}
    name: otc-container
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2023-02-13T01:17:14Z"
  hostIP: 10.0.4.233
  phase: Running
  podIP: 10.0.5.150
  podIPs:
  - ip: 10.0.5.150
  qosClass: BestEffort
  startTime: "2023-02-13T01:17:13Z"
```

**Comment by TylerHelmuth** [comment]
@sang-w0o I think this is going to end up being an issue with one of the python instrumentation libraries and not the operator.  Looking at the [python instrumentation libraries](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation) I believe the only one in your stack trace that python knows how to instrument is grpc.

What version of the operator are you using?  What versions of gRPC are you using?

**Comment by TylerHelmuth** [comment]
@sang-w0o I think this is going to end up being an issue with one of the python instrumentation libraries and not the operator.  Looking at the [python instrumentation libraries](https://github.com/open-telemetry/opentelemetry-python-contrib/tree/main/instrumentation) I believe the only one in your stack trace that python knows how to instrument is grpc.

What version of the operator are you using?  What versions of gRPC are you using?

**Comment by sang-w0o** [comment]
> What version of the operator are you using? What versions of gRPC are you using?

For operator, I just executed below command few days ago.

```sh
kubectl apply -f https://github.com/open-telemetry/opentelemetry-operator/releases/latest/download/opentelemetry-operator.yaml
```

For gRPC, we are just using the latest version without specifying version.
However, when I check `requirements.txt` of opentelemtry-operator, it seems that it provides `opentelemetry-instrumentation-grpc`, whereas I am using `grpcio`. Can this be an issue?

**Comment by TylerHelmuth** [comment]
It might be an issue. Id have to dig deeper into the instrumentation libraries. I don't think this is an operator issue, tho. @open-telemetry/operator-approvers @open-telemetry/opentelemetry-python-contrib-approvers can you move the issue to the contrib python repo?

**Comment by Aneurysm9** [comment]
I only have the option to move this to other OTel repos where I have write access, which doesn't include python-contrib.  You may need someone from the TC as they'll have write access in all repos.

**Comment by sang-w0o** [comment]
I was confused if this was an issue related to `Instrument` CR or python-instrumentation repo. 😵‍💫

**Comment by TylerHelmuth** [comment]
@sang-w0o opening the issue here isn't a problem since you're using the operator. With your added information I think we can now say this is an instrumentation issue. Thank you for bringing this to our attention 🙂

**Comment by srikanthccv** [comment]
@sang-w0o please create an issue here https://github.com/open-telemetry/opentelemetry-python-contrib/.

**Comment by pavolloffay** [comment]
@matej-g / @sang-w0o @TylerHelmuth   can be this issue closed now as there is https://github.com/open-telemetry/opentelemetry-python/issues/3175?

**Comment by matej-g** [comment]
I think there are two issues - one is with samplers, the other is that the application actually crashed. For the first one, I opened https://github.com/open-telemetry/opentelemetry-python/issues/3175, for the second one I'd recommend @sang-w0o to open an issue in the Python contrib. Sorry for the confusion.

I think there's nothing to be done in operator, so IMO this can be closed.

**Comment by sang-w0o** [comment]
Got it. Thanks for such kind support 🙏

**Comment by pl4nty** [comment]
Looks like this is fixed in [0.38b0](https://pypi.org/project/opentelemetry-instrumentation/0.38b0/) based on [my test image](https://github.com/pl4nty/containers/pkgs/container/opentelemetry-operator%2Fautoinstrumentation-python), so should be fixed once the official image is updated
